### PR TITLE
maestro record

### PR DESCRIPTION
**Pull request by Leland-Takamine**
`maestro record YourFlow.yaml` to render a video like the one below:

https://user-images.githubusercontent.com/847683/204396313-eee93a07-be32-4fa2-9292-e2285d44d762.mp4

Command line output:

https://user-images.githubusercontent.com/847683/204397309-829a24af-b55d-45d8-8d8b-fc850b4319a8.mov

### CODE REVIEW

File: maestro-cli/src/main/java/maestro/cli/api/ApiClient.kt [hunk 1]
```diff
@@ -130,6 +132,46 @@ class ApiClient(
         }
     }
 
+    fun render(
+        screenRecording: File,
+        frames: List<ResultView.Frame>,
+        progressListener: (totalBytes: Long, bytesWritten: Long) -> Unit = { _, _ -> },
+    ): String {
+        val baseUrl = "https://maestro-record.ngrok.io"
```
Comment (dmitry-zaitsev) [inline maestro-cli/src/main/java/maestro/cli/api/ApiClient.kt:140]: I suggest we setup a redirect somewhere on our end to be able to move away from ngrok later on

**Review by dmitry-zaitsev — APPROVED** [review]

File: maestro-cli/src/main/java/maestro/cli/command/RecordCommand.kt [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        val resultView = ResultView()
+        val screenRecording = kotlin.io.path.createTempFile(suffix = ".mp4").toFile()
+        val exitCode = screenRecording.sink().use { out ->
+            maestro.startScreenRecording(out).use {
+                TestRunner.runSingle(maestro, device, flowFile, env, resultView)
+            }
+        }
+
+        System.err.println()
+        System.err.println("@|bold Rendering your video. This usually takes a couple minutes...|@".render())
```
[the remark this answers is not in the document]
Comment (Leland-Takamine) [reply maestro-cli/src/main/java/maestro/cli/command/RecordCommand.kt:84]: Yeah this is useful when piping information between tools. eg. Just piping the download URL to another command

File: maestro-ios/src/main/java/ios/idb/IdbIOSDevice.kt [hunk 3]
```diff
@@ -418,6 +425,72 @@ class IdbIOSDevice(
         }
     }
 
+    // Warning: This method reads all bytes into memory. This can probably be optimized if necessary.
+    override fun startScreenRecording(out: Sink): Result<IOSScreenRecording, Throwable> {
+        val future = CompletableFuture<Void>()
```
Comment (amanjeetsingh150) [inline maestro-ios/src/main/java/ios/idb/IdbIOSDevice.kt:430]: Can we use xcrun simct recordVideo command here directly?

Comment (Leland-Takamine) [reply maestro-ios/src/main/java/ios/idb/IdbIOSDevice.kt:430]: What is the benefit of that?

File: maestro-cli/src/main/java/maestro/cli/command/RecordCommand.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        System.err.println()
+
+        val frames = resultView.getFrames()
+        val client = ApiClient("")
```
[the remark this answers is not in the document]
Comment (Leland-Takamine) [reply maestro-cli/src/main/java/maestro/cli/command/RecordCommand.kt:88]: This particular API call doesn't use the base url parameter

File: maestro-client/src/main/java/maestro/Maestro.kt [hunk 5]
```diff
@@ -361,6 +361,17 @@ class Maestro(private val driver: Driver) : AutoCloseable {
         }
     }
 
+    fun startScreenRecording(out: Sink): ScreenRecording {
+        LOGGER.info("Starting screen recording")
+        val screenRecording = driver.startScreenRecording(out)
+        return object : ScreenRecording {
+            override fun close() {
+                LOGGER.info("Stopping screen recording")
+                screenRecording.close()
```
[the remark this answers is not in the document]
Comment (Leland-Takamine) [reply maestro-client/src/main/java/maestro/Maestro.kt:370]: This should just work I believe but I did not test that. The service actually rejects any recording over 1 minute anyway

File: maestro-ios/src/main/java/ios/idb/IdbIOSDevice.kt (hunk 3 above)
Comment (amanjeetsingh150) [reply maestro-ios/src/main/java/ios/idb/IdbIOSDevice.kt:430]: Not depending on idb

Comment (Leland-Takamine) [reply maestro-ios/src/main/java/ios/idb/IdbIOSDevice.kt:430]: Let's address that holistically once we have a solution ready

**Pull request merged**